### PR TITLE
Add supportsBatch and supportsStreaming field to algorithms and component descriptors.

### PR DESCRIPTION
**Pull request by brosenberg42**
Issue #144

---
This change is [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/163)

### CODE REVIEW

**Comment by jrobble** [comment]
:lgtm:

---

Reviewed 3 of 4 files at r3.
Review status: 3 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/PipelineServiceImpl.java, line 609 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHUke1AirGhKP_asax:-KpHUke1AirGhKP_asay:bnc8n0f) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/PipelineServiceImpl.java#L609)):*
> ```Java
>     }
> 
>     /** Check that the task is not null, is valid, has a name which has not already been added to the PipelineManager, references only actions which have already been added to the PipelineManager, and if more than one action is specified, check that all referenced actions are of the same Operation. Returns false if the task is invalid. */
> ```

Update this comment.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/PipelineServiceImpl.java, line 696 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHUh4dLHARpGE6jmTZ:-KpHUh4dLHARpGE6jmT_:bnc8n0f) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/PipelineServiceImpl.java#L696)):*
> ```Java
>     }
> 
>     /** Check that the pipeline is not null, is valid, has a name which is unique in the PipelineManager, references valid tasks, and that the states (using provides/requires) are valid for the proposed sequence of tasks. Returns false if the pipeline. */
> ```

Update this comment.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/CustomPipelineValidatorImpl.java, line 383 at r1](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpGULKC2ml3v6JNpkpb:-KpGULKC2ml3v6JNpkpc:bdorhkp) ([raw file](https://github.com/openmpf/openmpf/blob/83db54d7421c3f6347c03b0291f188262eb9a2ab/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/CustomPipelineValidatorImpl.java#L383)):*
> ```Java
>         ActionDefinition action = pipelineService.getAction(actionName);
>         if (action == null) {
>             return EnumSet.allOf(ProcessingType.class);
> ```

Drop a comment to explain why the superset is returned.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java, line 125 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHX7Tmq7bLoH31QTvI:-KpHX7Tmq7bLoH31QTvJ:bauyy8i) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java#L125)):*
> ```Java
>         public boolean supportsBatchProcessing;
> 
>         public boolean supportsStreamProcessing;
> ```

One or both of these must be set for the JSON to be valid. We perform a similar kind of check here for `defaultValue` and `propertiesKey` by using `@ValidAlgoPropValue`:

```
    @ValidAlgoPropValue
    public static class AlgoProvidesProp {
        @NotBlank
        public String description;

        @NotBlank
        public String name;

        @NotNull
        public ValueType type;

        // option A
        public String defaultValue;

        // option B
        public String propertiesKey;
    }
```

I realize that `CustomPipelineValidatorImpl.validate()` somewhat covers this check, but it can be handled much sooner in the JSON validation itself.

---

*[trunk/workflow-manager/src/test/java/org/mitre/mpf/wfm/service/component/TestCustomPipelineValidator.java, line 259 at r1](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHRLx0qD7txO3EdUdO:-KpHRLx0qD7txO3EdUdP:b-1lutaa) ([raw file](https://github.com/openmpf/openmpf/blob/83db54d7421c3f6347c03b0291f188262eb9a2ab/trunk/workflow-manager/src/test/java/org/mitre/mpf/wfm/service/component/TestCustomPipelineValidator.java#L259)):*
> ```Java
>         descriptor.tasks.add(batchOnlyTask);
> 
>         JsonComponentDescriptor.Pipeline invalidPipeline = new JsonComponentDescriptor.Pipeline();
> ```

`Drop a comment saying this pipeline is invalid because the second action of the first task only supports streaming.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openmpf/openmpf/163#-:-KpHT8ShO2QKDCqOqf-3:bnfp4nl)*

**Comment by brosenberg42** [comment]
Review status: 0 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/PipelineServiceImpl.java, line 609 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHUke1AirGhKP_asax:-KpKeAviPzoH99iz3GES:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/PipelineServiceImpl.java#L609)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

Update this comment.
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/PipelineServiceImpl.java, line 696 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHUh4dLHARpGE6jmTZ:-KpKeBc3SxlMTyzDBhTf:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/PipelineServiceImpl.java#L696)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

Update this comment.
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/CustomPipelineValidatorImpl.java, line 383 at r1](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpGULKC2ml3v6JNpkpb:-KpKeEj4H-OSpgDEsl33:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/83db54d7421c3f6347c03b0291f188262eb9a2ab/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/CustomPipelineValidatorImpl.java#L383)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

Drop a comment to explain why the superset is returned.
</blockquote></details>

Done.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java, line 125 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHX7Tmq7bLoH31QTvI:-KpKeKKLBvri5rr79FrE:bxn55kd) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java#L125)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

One or both of these must be set for the JSON to be valid. We perform a similar kind of check here for `defaultValue` and `propertiesKey` by using `@ValidAlgoPropValue`:

```
    @ValidAlgoPropValue
    public static class AlgoProvidesProp {
        @NotBlank
        public String description;

        @NotBlank
        public String name;

        @NotNull
        public ValueType type;

        // option A
        public String defaultValue;

        // option B
        public String propertiesKey;
    }
```

I realize that `CustomPipelineValidatorImpl.validate()` somewhat covers this check, but it can be handled much sooner in the JSON validation itself.
</blockquote></details>

I prefer using the `@ScriptAssert` annotation to do this validation. The validation is a very simple expression, so it seems like overkill to create an annotation and validator to check this.

---

*[trunk/workflow-manager/src/test/java/org/mitre/mpf/wfm/service/component/TestCustomPipelineValidator.java, line 259 at r1](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHRLx0qD7txO3EdUdO:-KpKeGSymHsmlpg0IkrA:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/83db54d7421c3f6347c03b0291f188262eb9a2ab/trunk/workflow-manager/src/test/java/org/mitre/mpf/wfm/service/component/TestCustomPipelineValidator.java#L259)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

`Drop a comment saying this pipeline is invalid because the second action of the first task only supports streaming.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openmpf/openmpf/163)*

**Comment by brosenberg42** [comment]
*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java, line 125 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHX7Tmq7bLoH31QTvI:-KpKgzwy0AVJ3Xkw4iRN:be3yjda) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java#L125)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

I prefer using the `@ScriptAssert` annotation to do this validation. The validation is a very simple expression, so it seems like overkill to create an annotation and validator to check this.
</blockquote></details>

The `@ScriptAssert` validation occurs at the same time as the `@ValidAlgoPropValue` validation.
This is tested in TestDescriptorValidator lines 225 - 241

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openmpf/openmpf/163)*

**Comment by jrobble** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r4.
Review status: 3 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java, line 125 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHX7Tmq7bLoH31QTvI:-KpL42XLtlbkQ4JYvAId:b-f0o4n2) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java#L125)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

The `@ScriptAssert` validation occurs at the same time as the `@ValidAlgoPropValue` validation.
This is tested in TestDescriptorValidator lines 225 - 241
</blockquote></details>

I like @ScriptAssert better. Can you add one instead of @ValidAlgoPropValue for consistency?

---

*[trunk/workflow-manager/src/test/java/org/mitre/mpf/wfm/service/component/TestCustomPipelineValidator.java, line 259 at r1](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHRLx0qD7txO3EdUdO:-KpL3cAY_IBAj-s5JLkC:bsgo93z) ([raw file](https://github.com/openmpf/openmpf/blob/83db54d7421c3f6347c03b0291f188262eb9a2ab/trunk/workflow-manager/src/test/java/org/mitre/mpf/wfm/service/component/TestCustomPipelineValidator.java#L259)):*
> // Invalid because the first task's second action only supports streaming, making the task overall only support streaming.

Let's reword this to: 

// Invalid because the first task's second action only supports streaming, which would require every action in the pipeline to support streaming. This is not the case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openmpf/openmpf/163#-:-KpL4GLlS3Jbme7ElYfm:bnfp4nl)*

**Comment by brosenberg42** [comment]
*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java, line 125 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHX7Tmq7bLoH31QTvI:-KpLHcC2z-pig5x5UHGH:bq2oj7u) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java#L125)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

I like @ScriptAssert better. Can you add one instead of @ValidAlgoPropValue for consistency?
</blockquote></details>

That annotation would end up being:
```
    @ScriptAssert.List({
            @ScriptAssert(lang = "javascript",
                    script = "_this.defaultValue != null && _this.propertiesKey == null || _this.defaultValue == null && _this.propertiesKey != null",
                    message = "must provide either a defaultValue or propertiesKey, but not both."),

            @ScriptAssert(lang = "javascript",
                    script = "_this.defaultValue != null || _this.propertiesKey == null || !_this.propertiesKey.trim().isEmpty()",
                    message = "propertiesKey may not be empty")
    })
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openmpf/openmpf/163)*

**Comment by brosenberg42** [comment]
*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java, line 125 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHX7Tmq7bLoH31QTvI:-KpLLlEpQLd3cIZM2d1u:b-r3potn) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java#L125)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

That annotation would end up being:
```
    @ScriptAssert.List({
            @ScriptAssert(lang = "javascript",
                    script = "_this.defaultValue != null && _this.propertiesKey == null || _this.defaultValue == null && _this.propertiesKey != null",
                    message = "must provide either a defaultValue or propertiesKey, but not both."),

            @ScriptAssert(lang = "javascript",
                    script = "_this.defaultValue != null || _this.propertiesKey == null || !_this.propertiesKey.trim().isEmpty()",
                    message = "propertiesKey may not be empty")
    })
```
</blockquote></details>

Are you sure you want me to make that change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openmpf/openmpf/163)*

**Comment by jrobble** [comment]
*[trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java, line 125 at r3](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHX7Tmq7bLoH31QTvI:-KpLQwobdSIUUI-TCxl4:b-r98qsf) ([raw file](https://github.com/openmpf/openmpf/blob/f425bdb18465d2875cc3a02025113476eceac31d/trunk/workflow-manager/src/main/java/org/mitre/mpf/wfm/service/component/JsonComponentDescriptor.java#L125)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

Are you sure you want me to make that change?
</blockquote></details>

The logic is a little dense, so let's leave it as is. Thanks for showing me how it could be done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openmpf/openmpf/163)*

**Comment by jrobble** [comment]
Reviewed 12 of 16 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/openmpf/openmpf/163)*

**Comment by brosenberg42** [comment]
Review status: 19 of 20 files reviewed at latest revision, 1 unresolved discussion.

---

*[trunk/workflow-manager/src/test/java/org/mitre/mpf/wfm/service/component/TestCustomPipelineValidator.java, line 259 at r1](https://reviewable.io:443/reviews/openmpf/openmpf/163#-KpHRLx0qD7txO3EdUdO:-KpLTJU3gqNCKXJKVC2R:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/83db54d7421c3f6347c03b0291f188262eb9a2ab/trunk/workflow-manager/src/test/java/org/mitre/mpf/wfm/service/component/TestCustomPipelineValidator.java#L259)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

> // Invalid because the first task's second action only supports streaming, making the task overall only support streaming.

Let's reword this to: 

// Invalid because the first task's second action only supports streaming, which would require every action in the pipeline to support streaming. This is not the case.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openmpf/openmpf/163)*

**Comment by jrobble** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/openmpf/openmpf/163#-:-KpLflsBuTESx7DZp-2Z:bnfp4nl)*
